feat: configurable force compact policy

ellutionist opened the pull request.

PragmaTwice committed:
Refactor script commands (#1313)

ellutionist committed:
feat: configurable force compact policy

PragmaTwice committed:
Use std::shared_mutex instead of our own ReadWriteLock implementation (#1314)

torwig committed:
Return empty string instead of nil in HGETALL, HVALS, HRANGEBYLEX responses (#1315)

Close #1311

PragmaTwice committed:
Use inline specifier to replace extern global constants (#1316)

ellutionist committed:
delay the file creation time check

PragmaTwice committed:
Refactor task runner for more robust runtime state (#1318)

xiaobiaozhao commented on src/config/config.h:
```diff
@@ -131,6 +131,8 @@ struct Config {
   Cron compact_cron;
   Cron bgsave_cron;
   CompactionCheckerRange compaction_checker_range{-1, -1};
+  int64_t force_compact_file_age;
```
```suggestion
  uint64_t force_compact_file_age;
```

xiaobiaozhao commented on kvrocks.conf:
```diff
@@ -462,6 +462,16 @@ profiling-sample-record-threshold-ms 100
+# and its percentage of deleted keys is higher than 
+# "force-compact-file-min-deleted-percentage", it will be forcely compacted as well.
+
+# force-compact-file-age 172800
```
default open ?

torwig committed:
Minor refactoring of stc/storage/storage.cc (#1321)

Enhance error messages, tidy the code base, and a few functions are renamed.

PragmaTwice committed:
Replace list with other cache-friendly containers (#1322)

PragmaTwice committed:
Move unit tests for redis types to a new directory (#1324)

We move `tests/cppunit/t_*.cc` to `tests/cppunit/types/*.cc`

PragmaTwice committed:
Introduce TBB to refactor task runners (#1323)

We refactor the TaskRunner class using blocking queues (tbb::concurrent_bounded_queue) in oneTBB (https://github.com/oneapi-src/oneTBB).

oneTBB provides more utilities about threads which can be used in kvrocks, like spin locks, rw mutex (more efficient), concurrent map, set, vector, and queue, concurrent algorithms, task arena, task scheduler (which has some difference than task runners), and task graphs.

Co-authored-by: hulk <hulk.website@gmail.com>

git-hulk committed:
Check if the checkpoint seq is in the WAL boundary before using (#1320)

tisonkun committed:
Update README.md (#1326)

PragmaTwice committed:
Relax the condition in unit tests of TaskRunner (#1329)

tisonkun committed:
test: Remove testdb and compactdb dirs on exit (#1327)

Signed-off-by: tison <wander4096@gmail.com>
Co-authored-by: Twice <twice.mliu@gmail.com>

tisonkun committed:
build: package sources in a folder (#1332)

Signed-off-by: tison <wander4096@gmail.com>

PragmaTwice committed:
Avoid crash due to task runner start failed (#1331)

aleksraiden committed:
Bump RocksDB 8.0.0 (#1333)

A new version of RocksDB is arrived - the first release of the 8.х branch.

- Bug fixing (incl. bug with WAL restoring, Get/MultiGet with async_io, flushing, etc.)
- Remove deprecated file system functions
- Remove a lot of deprecated/obsolete statistics entity
- Added a subcode of Status::Corruption, Status::SubCode::kMergeOperatorFailed, for users to identify corruption failures originating in the merge operator
- Feature: a new MultiGetEntity API that enables batched wide-column point lookups

Full changelog: https://github.com/facebook/rocksdb/releases/tag/v8.0.0

ellutionist committed:
Add default and range specification for the new config item

torwig committed:
Fix clang-tidy warnings (#1334)

git-hulk committed:
Remove the test db dir after deleting the storage in CPP unittest (#1336)

PragmaTwice committed:
Add clang-tidy checking for unit tests (#1337)

git-hulk committed:
Use waitForImportState instead of the sleep in Go test case (#1339)

enjoy-binbin committed:
Keep feature descriptions consistent with documentation (#1340)

Recently we refined feature words, see apache/incubator-kvrocks-website#66.
This PR makes the key features in the README consistent with the docs,
plus removed some extra spaces and # characters.

dongdongwcpp committed:
Relax race condition between subkey compaction thread (#1335)

Co-authored-by: git-hulk <hulk.website@gmail.com>
Co-authored-by: Twice <twice.mliu@gmail.com>
Co-authored-by: tison <wander4096@gmail.com>

PragmaTwice committed:
Make Status::Code::cOK private to avoid access directly (#1341)

ellutionist committed:
add GetConfig to Storage

PragmaTwice committed:
Implement the new encoding with 64bit size and expire time in milliseconds (#1342)

See proposal in #1033, after this change, these old data can still be readable/writable, but all new data will be written via the new encoding if `ENABLE_NEW_ENCODING` is turned on (currently default OFF).

NOTE: Although the encoding now support 64bit size, maybe some place in code still use int32_t/int so it cannot work well in large number of items more than 32bit. We can fix them after this PR.

Co-authored-by: Yaroslav <torwigua@gmail.com>
Co-authored-by: hulk <hulk.website@gmail.com>

enjoy-binbin committed:
Add SUBSTR command as redis (#1347)

Co-authored-by: Twice <twice.mliu@gmail.com>

PragmaTwice committed:
Add some readability checks and fix all report in current codebase (#1348)

git-hulk committed:
Fix missing stream increment batch in cluster migration (#1345)

Currently, the cluster slot migration will split into two stages: send
snapshots and sync WAL if the snapshot was sent. For the stream, we only
parsed the KVs from the snapshot and missed the WAL part.

Co-authored-by: Twice <twice@apache.org>

ellutionist committed:
change config field type && enhance config description

enjoy-binbin committed:
Fix stream_metadata enum -Wint-in-bool-context warning (#1349)

There is a warning in the code:
```
/incubator-kvrocks/src/storage/batch_extractor.cc:89:50: warning: enum constant in boolean context [-Wint-in-bool-context]
   89 |       StreamMetadata stream_metadata(kRedisStream);
```

This was introduced in #1345

PragmaTwice committed:
Round expire time for seconds to improve precision for 32bit-common-field encoding (#1352)

PragmaTwice committed:
Unify time retrieving and remove srand call in metadata construction (#1351)

enjoy-binbin committed:
Add CLIENT INFO command as Redis (#1354)

This PR adds the CLIENT INFO command as Redis, this command
returns information about the current client connection,
the output format is the same as the CLIENT LIST command.

CLIENT INFO was added since Redis 6.2.0

git-hulk committed:
Merge branch 'unstable' into force_compact_policy
